### ISSUE TITLE
Unable to login after migrating from 1.8.x to 1.20.x

### ISSUE THREAD

**Issue by tarelda** [go-gitea/gitea#27846]
### Description

I am trying to migrate my instance from current deployment to TrueNAS kubernetes deployment. I exported my instance using gitea dump and imported into appropriate containers.
Firstly, I figured that I have an issue with password hashing algorithm and changed passwd_hash_algo to pbkdf2_v1 for my user.
After that I am still unable to login and gitea yields error 500. Logs attached. 

### Gitea Version

1.20.5

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

https://gist.github.com/tarelda/af6fe775e85ac44cccd113a50e710b10

### How are you running Gitea?

TrueNAS Scale App through its kubernetes deployment. 

### Database

PostgreSQL

**Comment by wxiaoguang** [comment]
> pbkdf2_v1

It is not a valid name. It should be `pbkdf2` in database IIRC.

**Comment by tarelda** [comment]
> > pbkdf2_v1
> 
> It is not a valid name. It should be `pbkdf2` in database IIRC.
It yields the same result:

```
2023-10-31 07:36:34.164018+00:002023/10/31 07:36:34 ...ers/web/auth/auth.go:224:SignInPost() [E] UserSignIn: pq: value too long for type character varying(10)
2023-10-31 07:36:34.168803+00:002023/10/31 07:36:34 ...eb/routing/logger.go:102:func1() [I] router: completed POST /user/login for 192.168.3.26:59484, 500 Internal Server Error in 28.3ms @ auth/auth.go:170(auth.SignInPost)

```
I also tried creating new admin user to access gitea, but CLI command fails also with the same error:

`Command error: CreateUser: pq: value too long for type character varying(10)
`


> ps: `dump/restore` has various bugs. You had better to dump the database by a database tool, then let Gitea upgrade the database by running a new version.

Unfortunately, previously I was running MariaDB, but I want to do the switch to Postgres, so dumping db directly is kinda non-option here. Regarding updating itself, I manually imported db dump into Postgres. Then copied all of the necessary files and let Gitea update itself on the first startup. 

I also ran gitea doctor recrete-table, which I assume is supposed to regenerate tables and command succeded with bringing db version to current.

**Comment by tarelda** [comment]
https://github.com/go-gitea/gitea/blob/a1c232cae3d8827691297d02d6f4ba980a805cd2/routers/web/auth/auth.go#L202-L224

This doesn't seem more helpful, but I believe it means that error originates from UserSignIn method. Postgres description indicates that this is caused by column that has type varchar(10). In case of table user it points towards two fields - salt and rands. Both are not affected by this method directly, but by authenticator. Authenticate. 

https://github.com/go-gitea/gitea/blob/16d15ce087cd4b9ddf91f6acff7eacac24e6aac5/services/auth/source/db/authenticate.go#L60-L67

It seems that call to user.SetPassword should be an issue, because it touches salt and also is probably called by CLI commands responsible for password change and user creation.

Seems on point, because 

https://github.com/go-gitea/gitea/blob/16d15ce087cd4b9ddf91f6acff7eacac24e6aac5/models/user/user.go#L506

16 characters can't fit into 10 characters. Thus it seems that database was not updated, because someone probably checked if salt can't fit into database table.
I checked my database dump and as you probably can already guess version table was missing. I checked original installation and version was 83. Then I updated current database version to 83 and ORM applied all of the migrations correctly. 

My suggestion is to update [docs](https://docs.gitea.com/administration/backup-and-restore#restore-command-restore) to have note **that user should check if database dump includes version table and its value is the same as in original database.** Otherwise migration will fail.
